face-detection-explainer.md

```diff
@@ -12,66 +12,125 @@
+* Face detection API does not need to return a mesh corresponding to the detected faces. Even though TensorFlow returns a 468-landmark FaceMesh and most DNNs can return something similar, mesh is not supported on any platforms presently, and for the sake of simplicity, it is excluded for now. However, in the long term it may be appropriate to extend the face detection API to be able to also return mesh-based face detection results. This is left for future work.
 
-## Using the face detection API
+## Face detection API
 
+### Metadata
 
-The API consists of two parts: first, the constraint `faceDetectionMode` to `getUserMedia()` or `applyConstrains()` is used to negotiate and enable the desired face detection mode. Often, camera drivers run already internally face detection for 3A, so enabling face detection might just make the results available to Web applications. The corresponding fields are added also to media track settings and capabilities. 
+The API consists of two parts: first, metadata which describes the faces available in video frames. The metadata could be also set by user by creating a new video frame (modifying video frame metadata in existing frames is not allowed by the specification). The metadata could be used by WebCodecs encoders to improve video encoding quality, for example by allocating more bits to face areas in frames. As the encoding algorithms are not specified in standards, also the exact way how the facial metadata is used is not specified.
 
-The second part of the API adds a new field `detectedFaces` of type sequence of `DetectedFace` into WebCodecs `VideoFrameMetadata` which provides information of the detected faces on the frame. In `DetectedFace`, the field `id` is used to track faces between frames: the same non-zero `id` in a face between different frames indicates that it is the same face. The value of zero means that tracking is not supported. `probability` is the probability that the returned face is in fact a face and not a false detection. This should be always greater than zero but less than one, since no perfect algorithm exists. If probability is not specified, user agent sets the field to zero.
+```js
+partial dictionary VideoFrameMetadata {
+  sequence<HumanFace> humanFaces;
+};
 
-The field `contour` provides an arbitrary number of points which enclose the face. Sophisticated implementations could provide a large number of points but initial implementations are expected to provide four contour points located in the corners of a rectangular region which describes the face bounding box. User agent is allowed to provide a minimum of one point, in which case the point should be located at the center of the face.
+dictionary HumanFace {
+  long              id;
+  float             probability;
+  DOMRectReadOnly   boundingBox;

```

Comment:
Can we make boundingBox required, or is there a case where we would know that there is a face in the image but we do not know where?

Reply:
I know see that we have a kind of union boudingBox or face landmark, so we cannot make it required.

Reply:
> we have a kind of union boudingBox or face landmark

No, if with union you mean that only either one would be valid. The intent is to allow both face `boundingBox` and face landmarks simultaneously. The `boundingBox` within the `HumanFace` dict would indicate bounding box for the whole face. Landmarks are part of human face and they can be set or not set independently of the whole face bounding box.

The reason why I did not mark `boundingBox` as `required` is the same as for the `centerPoint`: to anticipate extensions to `HumanFace` where we could have something like contour or bitmask for the face and not need bounding box anymore. If you think it is not a problem to now mark `boundingBox` as `required` and later to remove the keyword when dictionary is extended with new types of location, I am happy to add the keyword now. However, I am worried that if the member is now required, Web apps could break if the keyword is removed later and bounding box made optional. Probably this wouldn't happen though, as `humanFaceDetectionMode` would need to be set to something else than `bounding-box` too.

> is there a case where we would know that there is a face in the image but we do not know where

With this draft version, there wouldn't be such a case. However, it could be future extension: add a new `ObjectDetectionMode`  called `presence` which just indicates that a face exists in the frame without its location. Use case could be for example to prevent screensaver while user is in the front of his/her computer. We did have the `presence` mode in some earlier draft, but I removed it to simplify.

Reply:
Understood, and right, in the example there is no need to write centerPoint?.x because the metadata originates from a video track with humanFaceLandmarkDetectionMode == "center-point". 

I agree with the suggestion to document the invariants. I changed the PR as follows:

bounding-box:  This source generates metadata into {{VideoFrameMetadata}} related to human face detection including the bounding box information for each detected face. As an input, this is interpreted as a command to turn the setting of human face detection to bounding-box.

> But I would mark centerPoint as required
After changes needed to unblock [issue 79](https://github.com/w3c/mediacapture-extensions/issues/79), this doesn't look anymore feasible.